doc/glossary.rst

```diff
@@ -1855,25 +1855,51 @@ See concept :term:`sample property`.
         See :ref:`group_cv`.
 
     ``sample_weight``
-        A relative weight for each sample.  Intuitively, if all weights are
-        integers, a weighted model or score should be equivalent to that
-        calculated when repeating the sample the number of times specified in
-        the weight.  Weights may be specified as floats, so that sample weights
-        are usually equivalent up to a constant positive scaling factor.
-
-        .. FIXME: Is this interpretation always the case in practice? We have no common tests.
-
-        Some estimators, such as decision trees, support negative weights.
-
-        .. FIXME: This feature or its absence may not be tested or documented in many estimators.
-
-        This is not entirely the case where other parameters of the model
-        consider the number of samples in a region, as with ``min_samples`` in
-        :class:`cluster.DBSCAN`.  In this case, a count of samples becomes
-        to a sum of their weights.
-
-        In classification, sample weights can also be specified as a function
-        of class with the :term:`class_weight` estimator :term:`parameter`.
+        A weight for each data point. Intuitively, if all weights are integers,
+        using them in an estimator or a :term:`scorer` is like duplicating each
+        data point as many times as the weight value. Weights may also be
+        specified as floats: many estimators and scorers are invariant to a
+        rescaling of the weights by a constant positive factor but there are
+        exceptions.
```

Comment:
Sorry but I don't quite follow how "Weights may also be specified as floats" relates to "many estimators and scorers are invariant to a rescaling of the weights by a constant positive factor but there are exceptions".

Reply:
Because floating-point valued weights could be used to express relative weights as opposed to absolute (integer) repetitions if you divide integer weights by their sum across the training set, for instance. However, there are known case where this is not equivalent (and we are still not sure if it's a bug or a feature).

Reply:
I *think* I see what you are saying.
Integer weights could also be expressed by a float e.g., float 'relative' weights of [0.1, 0.2, 0.3], can be rescaled by a constant factor of 10 and would become integer weights [1, 2, 3] ?

I guess if weights were scale invariant, the above weights would effectively be the same.

I was recently reading this conversation: https://github.com/numpy/numpy/issues/8935#issuecomment-1398689411 (it's about weights specifically in quantiles) but it made me question if some interpretations of weight *is* sample size dependent (i.e. scale variant)..?

Sorry, this is probably not helpful in moving this forward.

Reply:
Yes, the point is that many scikit-learn estimators have a `fit` method that is naturally sample_weight-scale invariant, but there are exceptions (sometimes only for some combinations of constructor param values). We would need to conduct an exhaustive survey similarly to what we did to check the validity of the integer frequency semantics.

For estimators that are weight-scale invariant, the semantics of fractional weights (passed as floating point values) is easy to define: they are the same as the integer frequency weights because dividing or multiplying by an arbitrary constant that does not impact the results.

For estimators that are not weight scale invariant, the semantics of non-integer weights is not easy to define, but I don't want to make the glossary entry too complex.